gcsfs/core.py

```diff
@@ -1217,25 +1303,55 @@ def _open(
         )
 
     @classmethod
-    def split_path(cls, path):
+    def _split_path(cls, path, version_aware=False):
         """
         Normalise GCS path string into bucket and key.
 
         Parameters
         ----------
         path : string
             Input path, like `gcs://mybucket/path/to/file`.
-            Path is of the form: '[gs|gcs://]bucket[/key]'
+            Path is of the form: '[gs|gcs://]bucket[/key][?querystring][#fragment]'
+
+        GCS allows object generation (object version) to be specified in either
+        the URL fragment or the `generation` query parameter. When provided,
+        the fragment will take priority over the `generation` query paramenter.
 
         Returns
         -------
-            (bucket, key) tuple
+            (bucket, key, generation) tuple
         """
         path = cls._strip_protocol(path).lstrip("/")
         if "/" not in path:
-            return path, ""
-        else:
-            return path.split("/", 1)
+            return path, "", None
+        bucket, keypart = path.split("/", 1)
+        key = keypart
+        generation = None
+        if version_aware:
+            parts = urlsplit(keypart)
+            try:
+                if parts.fragment:
+                    generation = parts.fragment
+                elif parts.query:
+                    parsed = parse_qs(parts.query)
+                    if "generation" in parsed:
+                        generation = parsed["generation"][0]
+                # Sanity check whether this could be a valid generation ID. If
+                # it is not, assume that # or ? characters are supposed to be
+                # part of the object name.
+                if generation is not None:
+                    int(generation)
+                    key = parts.path
+            except ValueError:
+                generation = None
+        return (
```

Comment:
I wonder if we can foresee any problems of having this method not fit the superclass in what it returns - a 3-tuple instead of a 2-tuple

Reply:
OK, actually never mind - I see that s3fs follows the same pattern, and we differentiate between strip_protocol and split_path. All good.